module-system/sov-modules-api/src/transaction.rs

```diff
@@ -63,23 +96,181 @@ impl<C: Context> Transaction<C> {
 #[cfg(feature = "native")]
 impl<C: Context> Transaction<C> {
     /// New signed transaction.
-    pub fn new_signed_tx(priv_key: &C::PrivateKey, mut message: Vec<u8>, nonce: u64) -> Self {
+    pub fn new_signed_tx(
+        priv_key: &C::PrivateKey,
+        mut message: Vec<u8>,
+        chain_id: u64,
+        gas_tip: u64,
+        nonce: u64,
+    ) -> Self {
         // Since we own the message already, try to add the serialized nonce in-place.
         // This lets us avoid a copy if the message vec has at least 8 bytes of extra capacity.
-        let original_length = message.len();
-        message.extend_from_slice(&nonce.to_le_bytes());
+        let len = message.len();
+
+        // resizes once to avoid potential multiple realloc
+        message.resize(len + EXTEND_MESSAGE_LEN, 0);
+
+        message[len..len + 8].copy_from_slice(&chain_id.to_le_bytes());
+        message[len + 8..len + 16].copy_from_slice(&gas_tip.to_le_bytes());
+        message[len + 16..len + 24].copy_from_slice(&nonce.to_le_bytes());
 
         let pub_key = priv_key.pub_key();
         let signature = priv_key.sign(&message);
 
         // Don't forget to truncate the message back to its original length!
-        message.truncate(original_length);
+        message.truncate(len);
 
         Self {
             signature,
             runtime_msg: message,
             pub_key,
+            chain_id,
+            gas_tip,
             nonce,
         }
     }
 }
+
+impl<Tx> UnsignedTransaction<Tx> {
+    pub const fn new(tx: Tx, chain_id: u64, gas_tip: u64) -> Self {
+        Self {
+            tx,
+            chain_id,
+            gas_tip,
+        }
+    }
+}
+
+impl<Tx> borsh::BorshSerialize for UnsignedTransaction<Tx>
+where
+    Tx: borsh::BorshSerialize,
+{
+    fn serialize<W: io::Write>(&self, writer: &mut W) -> io::Result<()> {
+        self.tx.serialize(writer)?;
+        self.chain_id.serialize(writer)?;
+        self.gas_tip.serialize(writer)?;
+
+        Ok(())
+    }
+}
+
+impl<Tx> borsh::BorshDeserialize for UnsignedTransaction<Tx>
```

Comment:
Why do we need this implementation? Can's this be auto derived?

Reply:
`UnsignedTransaction` might live without a borsh serialization implementation, as in

https://github.com/Sovereign-Labs/sovereign-sdk/blob/3ae9b7b92c8123864bf7471bfb4c086da93fe3af/module-system/sov-cli/src/wallet_state.rs#L10-L20

Reply:
Not sue I understand. What would be the problem for deriving borsh for `UnsignedTransaction ` by a macro?

Reply:
Resolved on https://github.com/Sovereign-Labs/sovereign-sdk/pull/1213/commits/aa3e742e72b17e54f3bb8c3d95e34324ebaaec09